src/main/kotlin/org/jitsi/jibri/service/impl/FileRecordingJibriService.kt

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright @ 2018 Atlassian Pty Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.jitsi.jibri.service.impl
+
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import org.jitsi.jibri.CallParams
+import org.jitsi.jibri.capture.ffmpeg.FfmpegCapturer
+import org.jitsi.jibri.capture.ffmpeg.executor.impl.FFMPEG_RESTART_ATTEMPTS
+import org.jitsi.jibri.selenium.JibriSelenium
+import org.jitsi.jibri.selenium.JibriSeleniumOptions
+import org.jitsi.jibri.service.JibriService
+import org.jitsi.jibri.service.JibriServiceStatus
+import org.jitsi.jibri.sink.Sink
+import org.jitsi.jibri.sink.impl.FileSink
+import org.jitsi.jibri.util.NameableThreadFactory
+import org.jitsi.jibri.util.ProcessMonitor
+import org.jitsi.jibri.util.WritableDirectory
+import org.jitsi.jibri.util.extensions.error
+import java.io.File
+import java.io.IOException
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import java.util.logging.Logger
+
+data class RecordingOptions(
+    /**
+     * The directory in which recordings should be created
+     */
+    val recordingDirectory: WritableDirectory,
+    /**
+     * The params needed to join the call
+     */
+    val callParams: CallParams,
+    /**
+     * The filesystem path to the script which should be executed when
+     *  the recording is finished.
+     */
+    val finalizeScriptPath: String
+)
+
+/**
+ * Set of metadata we'll put alongside the recording file(s)
+ */
+data class RecordingMetadata(
+    val participants: List<Map<String, Any>>
+)
+
+/**
+ * [FileRecordingJibriService] is the [JibriService] responsible for joining
+ * a web call, capturing its audio and video, and writing that audio and video
+ * to a file to be replayed later.
+ */
+class FileRecordingJibriService(private val recordingOptions: RecordingOptions) : JibriService() {
+    /**
+     * The [Logger] for this class
+     */
+    private val logger = Logger.getLogger(this::class.qualifiedName)
+    /**
+     * Used for the selenium interaction
+     */
+    private val jibriSelenium = JibriSelenium(
+        JibriSeleniumOptions(recordingOptions.callParams),
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("JibriSelenium"))
```

Comment:
Perhaps we can move this executor instantiation to a default value in JibriSelenium (to keep the string reference local)?
I'm curious why for the streaming service JibriSelenium shares the service's executor, but for the file recording it has its own?

Reply:
yeah, this stuff is a bit inconsistent.  executors were moved to be params in places as unit tests were written so that a mock could be passed in.  in this place i think i just happened to optimize a bit by having the service shared.  i need a good sweep on all the usages here to make them consistent.